nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/CopyS3Object.java

```diff
@@ -119,28 +145,127 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
         final String destinationBucket = context.getProperty(DESTINATION_BUCKET).evaluateAttributeExpressions(flowFile).getValue();
         final String destinationKey = context.getProperty(DESTINATION_KEY).evaluateAttributeExpressions(flowFile).getValue();
 
+        final GetObjectMetadataRequest sourceMetadataRequest = new GetObjectMetadataRequest(sourceBucket, sourceKey);
+        final ObjectMetadata metadataResult = s3.getObjectMetadata(sourceMetadataRequest);
+
+        if (metadataResult == null) {
+            session.transfer(flowFile, REL_FAILURE);
+        }
+
+        final AtomicReference<String> multipartIdRef = new AtomicReference<>();
+        final long contentLength = metadataResult.getContentLength();
+        final boolean isMultiPart = metadataResult.getContentLength() > MULTIPART_THRESHOLD;
+
         try {
-            final CopyObjectRequest request = new CopyObjectRequest(sourceBucket, sourceKey, destinationBucket, destinationKey);
             final AccessControlList acl = createACL(context, flowFile);
-            if (acl != null) {
-                request.setAccessControlList(acl);
-            }
-
             final CannedAccessControlList cannedAccessControlList = createCannedACL(context, flowFile);
-            if (cannedAccessControlList != null) {
-                request.setCannedAccessControlList(cannedAccessControlList);
-            }
 
-            s3.copyObject(request);
+            if (!isMultiPart) {
+                smallFileCopy(s3, acl, cannedAccessControlList, sourceBucket, sourceKey, destinationBucket, destinationKey);
+            } else {
+                final int retryLimit = context.getProperty(MULTIPART_RETRIES).asInteger();
+                largeFileCopy(s3, acl, cannedAccessControlList, sourceBucket, sourceKey, destinationBucket,
+                        destinationKey, multipartIdRef, contentLength, retryLimit);
+            }
             session.getProvenanceReporter().send(flowFile, getTransitUrl(destinationBucket, destinationKey));
             session.transfer(flowFile, REL_SUCCESS);
-        } catch (final IllegalArgumentException | AmazonClientException e) {
+        } catch (final ProcessException | IllegalArgumentException | AmazonClientException e) {
+            if (isMultiPart && !StringUtils.isEmpty(multipartIdRef.get())) {
+                try {
+                    AbortMultipartUploadRequest abortRequest = new AbortMultipartUploadRequest(destinationBucket, destinationKey, multipartIdRef.get());
+                    s3.abortMultipartUpload(abortRequest);
+                } catch (AmazonS3Exception ignored) {
+                    getLogger().error("Failed to cleanup the partial upload to bucket {} and key {}", destinationBucket, destinationKey);
+                    getLogger().error("Abort exception", ignored);
```

Comment:
This should be consolidated, also recommend changing this to a warning:
```suggestion
                } catch (final AmazonS3Exception s3e) {
                    getLogger().warn("Abort Multipart Upload failed for Bucket [{}] Key [{}]", destinationBucket, destinationKey, s3e);